### ISSUE TITLE
Jarring edges between underground biomes

### ISSUE THREAD

**Issue by tomfulghum** [yungnickyoung/Better-Caves#6]
I get some really pronounced (and in my opinion quite ugly) walls at the edges between underground biomes, most commonly where a cave biome meets a large cavern biome.

Pictures:
![2019-10-01_01 52 31](https://user-images.githubusercontent.com/5806810/65924850-d6625180-e3ee-11e9-8a1d-afa00254713b.png)
![2019-10-01_01 52 54](https://user-images.githubusercontent.com/5806810/65924852-d6fae800-e3ee-11e9-9855-0f97568c59d0.png)
![2019-10-01_01 53 02](https://user-images.githubusercontent.com/5806810/65924853-d6fae800-e3ee-11e9-904f-1341ce179d2d.png)
![2019-10-01_01 53 24](https://user-images.githubusercontent.com/5806810/65924854-d6625180-e3ee-11e9-858b-dd69e7623296.png)

The world seed is 7462823440804153335 if you want to check it out yourself.

Is this a bug and if not, is there a way to prevent this?

**Comment by moonsquig** [comment]
Just commenting to say that I am also experiencing these types of issues as I commented on the curseforge page. 

![image](https://user-images.githubusercontent.com/56037441/65975137-157fb980-e466-11e9-9515-5940e4d05fdd.png)

It's not terrible for me and I find that upping the size of the cave biomes reduces the chance of them occurring, but I still find them on occasion and they can be quite jarring.

**Comment by yungnickyoung** [comment]
Hey tomfulghum, sorry to hear you don't like this. This behavior is, in fact, intended. It's a consequence of the algorithm I've devised for this mod. The mod separates cave/cavern types based on randomly generated [Voronoi regions](https://i.imgur.com/fem3bJp.png). As such, the boundary between cave/cavern biomes will be distinct. I personally like the look of it, but I can understand why you wouldn't.

There isn't too much to be done to prevent this, but there are some options you can experiment with to help reduce these walls. Take a look at the Cavern Biome Size option as well as the Cavern Frequency settings for each of the caverns. Upping the Cavern Biome Size will make the caverns bigger, and therefore the walls will be further away and less frequently encountered. If you don't want the caverns at all, you can set each of the Cavern frequencies to None. Alternatively, set the Cavern Frequencies to VeryCommon and the Cavern Biome Size to ExtraLarge if you want almost all caverns at the bottom (which would also prevent walls).

Since this is intended behavior, I'm going to close this issue, but feel free to continue to give your comments, criticism, and ideas here.

moonsquig - While the walls are supposed to be there, it looks like there is a jarring right angle in the middle of your picture. Such square-ish corners should not exist. If you could make a separate issue with additional pictures and your complete modpack list, that would be great.

**Comment by tomfulghum** [comment]
Could smoothing out your Voronoi regions ([like this, for example](http://www.iquilezles.org/www/articles/smoothvoronoi/smoothvoronoi.htm)) and interpolating between the different biomes work?

Off Topic (mostly): If I understand your code correctly (and please correct me if I'm wrong), you're actually using the default (Simplex) noise for choosing the biome type, since you never actually set the noise type of the FastNoise instances/biome controllers, only their cellular return type, which has no effect if the noise type is not set to Cellular.

**Comment by yungnickyoung** [comment]
Holy shit I think you're right. I can't believe I didn't notice this LMAO thanks so much for pointing it out. This would explain a few things I've noticed in the past when testing. I'm surprised simplex noise works as well as it does for sectioning off biomes. 

Well, even though this hasn't really created any problems, I'll be messing around with the _intended_ noise type now, lol. Again, thanks so much for pointing this out! I feel like such an idiot for not noticing.

The smooth Voronoi is interesting, and a great idea! I'll look into the kinds of transitions I can create between regions.

**Comment by yungnickyoung** [comment]
So, I've decided to keep the simplex noise for the cavern biome controller, as I really like the way it can create caverns of various shapes. I went ahead and switched the cave biome controller and water biome controller to cellular noise.

The cavern biome controller now has smoothing along boundaries to eliminate almost all of the behavior exemplified by your images. This is done quite simply, thanks to the use of simplex noise for the controller. I pad the usual threshold for determining a given type of cavern with a "smoothing range." For example, if the threshold for floored caverns is 0.6 and above, I say its smoothing range is 0.45 to 0.6. Because of the continuous nature of simplex noise, this range is guaranteed to essentially form a ring around the cavern. If the noise is within this range for a given block, I determine how far into this range it is, and then convert that to a multiplier to reduce the noise threshold when mining blocks. Sounds complicated and kinda hard to explain, but check the code if you'd like - it's pretty straightforward.

Here's a little before and after:
Before:
![2019-10-03_23 18 19](https://user-images.githubusercontent.com/25270754/66179910-be7e0e00-e638-11e9-858f-2f61a37bcd23.png)
After:
![2019-10-03_23 18 22](https://user-images.githubusercontent.com/25270754/66179914-c3db5880-e638-11e9-8e3d-91f8936fb526.png)

Before:
![2019-10-03_23 19 11](https://user-images.githubusercontent.com/25270754/66179919-cb026680-e638-11e9-8c79-105ce336ac23.png)
After:
![2019-10-03_23 19 03](https://user-images.githubusercontent.com/25270754/66179922-cf2e8400-e638-11e9-93ea-cb5254145fee.png)


This does add a performance hit, but it doesn't seem to be too significant. 
All of this will be available in v1.5 (uploading very soon). Hopefully you find it better than before!
